A32-xds-circuit-breaking.md

```diff
@@ -0,0 +1,147 @@
+picker will do the pick as usual, returning a subchannel. If not, it will 
+fail the call with status UNAVAILABLE. Such failed calls will not be retried,
+but they will be recorded to the `total_dropped_requests` counts in 
+cluster-level load reports and reported to the load reporting server.
+
+Note an EDS LB policy may be responsible for handling `service_name` switch
+used for endpoint discovery, requests sent to both the old and new services
+should be aggregated to the same counter and the total is restricted by the
+configured limit. Even though the failed requests will be recorded and reported
+in separate stats message.
+
+## Implementation
+
+### Java
+- Add a `max_requests` field in the EDS LB config. The value will be taken
+from the cluster watch interface pushed by the XdsClient.
+- Maintain a counter in the EDS LB policy for counting the number of 
+outstanding requests.
+- Wrap the `SubchannelPicker` propagated from each of EDS policy's child LB 
+policy with the logic of checking the counter and make pick decision.
+- Counter manipulation is done in `ClientStreamTracer`/`Factory`: increment
+in `newClientStreamTracer()` and decrement in `streamClosed()`.
+    - Note there will be check-and-allocate race as the pick method and stream
+    tracer are not run by the same thread. This will cause the limit to
+    be exceeded. Since the race window is small, the value should be exceeded 
+    only by a small amount.
+
+### C
+TODO
+
+### Go
+TODO
+
+## Other parameters considered
+
+There is a group of other parameters in the xDS circuit breakers, but most of
+them do not apply to gRPC's use case, and they will be ignored in the 
+implementation. The following parameters have been considered:
+
+- `max_connections`: the maximum number of connections can be created to an 
+upstream cluster.
+    - Envoy allows at least one connection to be allocated for each host 
+    selected by cluster load balancing, even if this circuit breaker is
+    overflowed. On the other hand, gRPC creates at most one connection to each
+    host. So this parameter does not have effective meaning in gRPC's use
+    case.
+- `max_pending_requests`: the maximum number of requests that will be queued
+while waiting for a connection to be established.
+    - As a proxy, Envoy drops queued requests from downstream entities to avoid
+    unbounded congestion and memory usage of the process. But as part of the
+    application, it makes little sense for gRPC to drop its own requests.
+- `max_retries`: retry related parameter, the retry feature has not been 
+implemented in gRPC xDS.
+- `max_budget`: retry related parameter, the retry feature has not been 
+implemented in gRPC xDS.
+- `max_connection_pools`: gRPC does not use connection pooling.
+- `connect_timeout` (in `Cluster` configuration): hard to implement in gRPC's 
+architecture, will reconsider after getting usage feedback.
+- `max_requests_per_connection` (in `Cluster` configuration): lifetime limit 
```

Comment:
I think there is general value in this one, since it allows you to put in place anti-hotspotting mitigations, i.e. forcing new connections after a certain amount of traffic. But, this can also be done with connection lifetime timeouts.

Reply:
> forcing new connections after a certain amount of traffic

Since we will create at most one connection per endpoint, this behavior would be shutting down the connection and re-establish. This doesn't seem to do anything helpful. Also we are not able to have a timely connection shutdown and re-establishment, we still need to let ongoing RPCs continue and finish. So during the period after the threshold is hit, the effective behavior would be the same as this Subchannel has been shut down. And it will be recreated again. The effective observation is the endpoint is _temporarily_ unreachable by new RPCs, and how long it will be unreachable depends on when ongoing RPCs sending to it finishes. This sounds to be a very weird thing.

Reply:
I agree that if you have this restriction of maximum of one connection, it's hard to sensibly implement this policy without blocking the channel during drain.